kmk/encoder.py

```diff
@@ -0,0 +1,151 @@
+import digitalio
+from .kmktime import ticks_ms
+
+# TODO: add velocity for internal state change
+
+class Encoder:
+    def __init__(
+        self,
+        pad_a,
+        pad_b,
+        button_pin=None,
+        is_inverted=False,
+        increment_key=None, # this is available but not  preferred
+        decrement_key=None, # this is available but not  preferred
+        vel_mode=False,
+        use_map = False # this is the preferred way, example in pytreus key map
+    ):
+        self.pad_a = self.PreparePin(pad_a)  # board pin for enc pin a
+        self.pad_b = self.PreparePin(pad_b)  # board pin for enc pin b
+        self.button_pin = self.PreparePin(button_pin)  # board pin for enc btn
+        self.button_state = None  # state of pushbutton on encoder if enabled
+        self.encoder_value = 0   # clarify what this value is
+        self.encoder_state = "00"  # quaderature encoder state
+        self.encoder_direction = None  # arbitrary, tells direction of knob
+        self.last_encoder_state = None  # not used yet
+        self.resolution = 2  # number of keys sent per position change
+        self.revolution_count = 20  # position changes per revolution
+        self.has_button = False  # enable/disable button functionality
+        self.encoder_data = None  # 6tuple containing all encoder data
+        self.position_change = None  # revolution count, inc/dec as knob turns
+        self.last_encoder_value = 0  # not used
+        self.is_inverted = is_inverted  # switch to invert knob direction
+        self.increment_key = increment_key  # list of increment keys
+        self.decrement_key = decrement_key  # list of decrement keys
+        self.vel_mode = vel_mode  # enable the velocity ouput
+        self.vel_ts = None  # velocity timestamp
+        self.last_vel_ts = 0  # last velocity timestamp
+        self.debug = False  # do not spew debug info by default
+        self.encoder_speed = None  # ms per position change(4 states)
+        self.use_map = use_map # look for encoder map
+
+    # adapted for CircuitPython from raspi
+    def PreparePin(self, num):
+        if num is not None:
+            pad = digitalio.DigitalInOut(num)
+            pad.direction = digitalio.Direction.INPUT
+            pad.pull = digitalio.Pull.UP
+            return pad
+        else:
+            return None
+
+    # checks encoder pins, reports encoder data
+    # logic is from https://github.com/nstansby/rpi-rotary-encoder-python/blob/master/encoder.py
+    def report(self):
+        encoder_pad_a = int(self.pad_a.value)
+        encoder_pad_b = int(self.pad_b.value)
+        new_encoder_state = "{}{}".format(encoder_pad_a, encoder_pad_b)
+
+        if self.encoder_state == "11":  # Resting position
```

Comment:
While there's some examples of "stringly-typed" code in the system still that we haven't bothered to rip out, I broadly find that [enums are a bit cleaner to work with](https://github.com/KMKfw/kmk_firmware/blob/6baaf5e5d430c82f6f01cde41875a943b247550e/kmk/consts.py#L7), and while I'm not sure entirely, I have a hunch that they're more performant on microcontrollers than strings, too

Reply:
They are, especially when used with const types. Enums should be preferred, and set to const where possible.

Reply:
to be more constructive here, a potential alternative implementation here then would be:

```python
class EncoderPadState:
    OFF = False
    ON = True

# Convenience for reverse lookup
EncoderPadLookup = {
    False: EncoderPadState.OFF,
    True: EncoderPadState.ON,
}

# ...

new_encoder_state = (EncoderPadLookup[self.pad_a.value], EncoderPadLookup[self.pad_b.value])
```

This gives you a tuple of (A, B) with, ultimately, True/False values, but more human readably so, because we can then compare `new_encoder_state[0]` against `EncoderPadState.OFF`, which gives a bit clearer of a picture of what's going on to a reader.

There's further renditions of this that involve [namedtuples](https://circuitpython.readthedocs.io/en/latest/docs/library/collections.html#collections.namedtuple) to make things even more readable, but that's both optional, and not guaranteed to be portable across MCUs since `collections` is stripped from some CPy builds